### ISSUE TITLE
Warning when constructor name is used as variable in match

### ISSUE THREAD

**Issue by RustanLeino** [dafny-lang/dafny#1600]
Consider the following program, which contains a simple mistake:

``` dafny
datatype Color = Red | Green | ShadesOfGray(nat)

predicate Monochromatic(c: Color) {
  match c
  case ShadesOfGray => true
  case anythingElse => false
}
```

For this program, Dafny issues a "redundant branch" warning for the second `case`. Why? Because the _constructor_ `ShadesOfGray` takes a parameter, and since the _pattern_ `ShadesOfGray` does not list a parameter, Dafny interprets the `ShadesOfGray` in the first pattern as a variable (which will match any `c`, just like the variable `anythingElse` in the second pattern would match anything that does not match in a prior `case`).

It would be nice if Dafny issued a warning in this case. I can imagine

    Warning: "ShadesOfGray" on line 5 is used as a variable, but it is also the name of a constructor that can be written here. If you intended to write a pattern that uses this constructor, add the parameters, like `ShadesOfGray(_)`.

Such a warning could be emitted once all resolution/type inference/type checking has been done. The check would then traverse through all `case` patterns. If inside a pattern it finds a variable `X` of type `T`, where `T` is a datatype that contains a constructor (with 1 or more parameters) named `X`, then the warning above is generated. Alternatively, this check could be implemented earlier, at the point where the Resolver today distinguishes between a nullary constructor (like `Red`) and a variable.

**Comment by seebees** [comment]
I agree. My issue with "redundant branch" wording is that it makes me think that I can clean something up because these branches "do the same thing". But as you point out, in this case Dafny means "I will never go into that branch".
If I have 2 copies of the same function that do the same thing, one of them is redundant.
However, I would never expect that callsites for the redundant function would never execute.
Not exactly the same example, I know. But it was how I was interpreting that warning.

**Comment by cpitclaudel** [comment]
Oh, it looks like my recent issue https://github.com/dafny-lang/dafny/issues/1789 is a duplicate of this one.

**Comment by robin-aws** [comment]
I thought of another way we could attack this. The code sample should also produce a style warning something along the lines of:

```
Warning: variables should begin with a lowercase letter: "ShadesOfGray" 
```

That's a lot shorter than trying to explain the specific situation.

**Comment by keyboardDrummer** [comment]
A related but slightly different issue: https://github.com/dafny-lang/dafny/issues/1991

**Comment by kjx** [comment]
this is still biting me 5 years later, for what it's worth. 
I get warnings that the "Rat" and "Cat" lets of the match are redundant, but that's it. and the code still verifies and runs
"woof woof woof". 


```
datatype Pet =
  | Dog(breed : string)
  | Rat
  | Cat

method talk(p : Pet)
  {
    match p
      case Dog => print p.breed, "woof ";
      case Rat => print "eeep ";
      case Cat => print "meow ";
  }

method Main()
  {
    talk(Rat);
    talk(Cat);
    talk(Dog("hound"));
    print "\n";
  }
```

I'm coming around to the Haskell-like ideas of enforced upper vs lower case distinction (although as a colleague points out, that works only in Latin scripts).    But I also like the syntactic parallelism here, and because Dafny as OO dot-syntax destructors, there's no need to repeat the datatype's fields in every match/case leg.   

Plus, there the "Dog" case should (but currently does not) flow-type the "selector" down to the type of each match/case leg. so I can do things like this: 

```
    match p
      case Dog => print p.breed, "woof ";
```
